main.py

```diff
@@ -51,58 +55,57 @@ def update_colors(self, links):
-                                              'flows': {}}
-            else:
-                self.switches[switch.dpid]['neighbors'] = set()
-
-        for link in links:
-            source = link['endpoint_a']['switch']
-            target = link['endpoint_b']['switch']
-            if source != target:
-                self.switches[source]['neighbors'].add(target)
-                self.switches[target]['neighbors'].add(source)
+        with self._switches_lock:
+            for switch in self.controller.switches.copy().values():
+                if switch.dpid not in self.switches:
+                    color = int(switch.dpid.replace(':', '')[4:], 16)
+                    self.switches[switch.dpid] = {'color': color,
+                                                  'neighbors': set(),
+                                                  'flows': {}}
+                else:
+                    self.switches[switch.dpid]['neighbors'] = set()
+
+            for link in links:
+                source = link['endpoint_a']['switch']
+                target = link['endpoint_b']['switch']
+                if source != target:
+                    self.switches[source]['neighbors'].add(target)
+                    self.switches[target]['neighbors'].add(source)
+
+        dpid_flows = defaultdict(list)
 
         # Create the flows for each neighbor of each switch and installs it
         # if not already installed
-        for dpid, switch_dict in self.switches.items():
-            switch = self.controller.get_switch_by_dpid(dpid)
-            if switch.ofp_version == '0x01':
-                controller_port = Port.OFPP_CONTROLLER
-            elif switch.ofp_version == '0x04':
-                controller_port = PortNo.OFPP_CONTROLLER
-            else:
-                continue
-            for neighbor in switch_dict['neighbors']:
-                if neighbor not in switch_dict['flows']:
-                    flow_dict = {
-                        'table_id': 0,
-                        'match': {},
-                        'priority': 50000,
-                        'actions': [
-                            {'action_type': 'output', 'port': controller_port}
-                        ]}
-
-                    flow_dict['match'][settings.COLOR_FIELD] = \
-                        self.color_to_field(
-                            self.switches[neighbor]['color'],
-                            settings.COLOR_FIELD
-                        )
-
-                    switch_dict['flows'][neighbor] = flow_dict
-                    returned = requests.post(
-                        url % dpid,
-                        json={'flows': [flow_dict]}
-                    )
-                    if returned.status_code // 100 != 2:
-                        log.error('Flow manager returned an error inserting '
-                                  f'flow. Status code {returned.status_code}')
+        with self._switches_lock:
```

Comment:
The current rw accesses on `self.switches` made it a bit harder to only solve with a shallow copy, it's also updating nested dict entries on a threaded handler so a lock here is the safest option for now, in the future, we can assess other possibilities here, either by refactoring more the data structure or using an async handler to avoid preemption at certain points.